modbus: serial: Allow custom stop-bit settings in server mode

msalau opened the pull request:
A Modbus server should be able to operate with the same settings as a client. Stop-bits -is the only setting not available in server mode.

The change makes it possible to override default stop-bits settings in backwards compatible way. The feature is disabled by default.

Commit:
modbus: serial: Add non-compliant mode with custom stop-bit settings

The mode is activated by the CONFIG_MODBUS_NONCOMPLIANT_SERIAL_MODE option
and allows any stop-bit setting for the serial port.

Signed-off-by: Maksim Salau <msalau@iotecha.com>

msalau commented on subsys/modbus/Kconfig:
```diff
@@ -79,6 +79,14 @@ config MODBUS_FC08_DIAGNOSTIC
 	help
 	  Enable function code 08 Diagnostic support
 
+config MODBUS_STOP_BITS_IN_SERVER_MODE
```
Hello @jfischer-no 
Thank you for feedback!
Yes, I'm fully aware that 8N1 is not compliant with the specification but there are devices on the market that use exactly this format.
Right now the Modbus subsystem can talk to a non-compliant server device as a client but can't act as a non-compliant server. So we can't create a server device that may be added to an existing non-compliant bus.
This is exactly my case: an existing bus that works at 9600 8N1 and I need to add another server to it without changing the communication format.

May be a better description of the config option is needed to highlight that this options is not compliant with the specification and should be used with caution.

jfischer-no replied:
understood, I am fine with
```
config MODBUS_NONCOMPLIANT_RTU_MODE
	bool "Non-compliant RTU server mode"
	depends on MODBUS_SERVER && MODBUS_SERIAL
	help
	  Allow non-compliant stop and parity bit settings.
```

jfischer-no commented on subsys/modbus/Kconfig:
```diff
@@ -79,6 +79,14 @@ config MODBUS_FC08_DIAGNOSTIC
 
+config MODBUS_STOP_BITS_IN_SERVER_MODE
+	bool "Use stop_bits_client parameter in server mode"
+	depends on MODBUS_SERVER
+	depends on MODBUS_SERIAL
```
`	depends on MODBUS_SERVER && MODBUS_SERIAL`

msalau commented:
Hello @jfischer-no 
Thank you for feedback. I've updated the change according to your comment.
I've renamed the option from `MODBUS_NONCOMPLIANT_RTU_MODE` to `MODBUS_NONCOMPLIANT_SERIAL_SERVER_MODE` since it affects not just RTU but ASCII also.

What about renaming `struct modbus_serial_param::stop_bits_client` into `stop_bits` and making non-compliant stop-bit configuration disabled by default for both clients and servers?
The current implementation allows non-compliant stop-bits configuration at all times for clients.

I.e.:
`modbus/Kconfig:`
```
config MODBUS_NONCOMPLIANT_SERIAL_MODE
	bool "Non-compliant serial mode"
	depends on MODBUS_SERIAL
	help
	  Allow non-compliant stop and parity bit settings.
```

`modbus/modbus.h`:
```diff
/**
 * @brief Modbus serial line parameter
 */
struct modbus_serial_param {
	/** Baudrate of the serial line */
	uint32_t baud;
	/** parity UART's parity setting:
	 *    UART_CFG_PARITY_NONE,
	 *    UART_CFG_PARITY_EVEN,
	 *    UART_CFG_PARITY_ODD
	 */
	enum uart_config_parity parity;
	/** stop_bits_client UART's stop bits setting if in client mode:
	 *    UART_CFG_STOP_BITS_0_5,
	 *    UART_CFG_STOP_BITS_1,
	 *    UART_CFG_STOP_BITS_1_5,
	 *    UART_CFG_STOP_BITS_2,
	 */
-	enum uart_config_stop_bits stop_bits_client;
+	enum uart_config_stop_bits stop_bits;
};
```

`modbus/modbus_serial.c`:
```diff
-	if (ctx->client || IS_ENABLED(MODBUS_NONCOMPLIANT_SERIAL_SERVER_MODE)) {
+ 	if (IS_ENABLED(MODBUS_NONCOMPLIANT_SERIAL_MODE)) {
		/* Allow custom stop bit settings */
		switch (param->serial.stop_bits_client) {
		case UART_CFG_STOP_BITS_0_5:
		case UART_CFG_STOP_BITS_1:
		case UART_CFG_STOP_BITS_1_5:
		case UART_CFG_STOP_BITS_2:
			uart_cfg.stop_bits = param->serial.stop_bits;
			break;
		default:
			return -EINVAL;
		}
	}
```

This will require some migration effort for existing applications but will explicitly enable non-compliant mode for both clients and servers.

jfischer-no commented:
> Hello @jfischer-no Thank you for feedback. I've updated the change according to your comment. I've renamed the option from `MODBUS_NONCOMPLIANT_RTU_MODE` to `MODBUS_NONCOMPLIANT_SERIAL_SERVER_MODE` since it affects not just RTU but ASCII also.
> 
> What about renaming `struct modbus_serial_param::stop_bits_client` into `stop_bits` and making non-compliant stop-bit configuration disabled by default for both clients and servers? The current implementation allows non-compliant stop-bits configuration at all times for clients.
> 
> I.e.: `modbus/Kconfig:`
> 
> ```
> config MODBUS_NONCOMPLIANT_SERIAL_MODE
> 	bool "Non-compliant serial mode"
> 	depends on MODBUS_SERIAL
> 	help
> 	  Allow non-compliant stop and parity bit settings.
> ```

Yes, that is event better.

jfischer-no reviewed (approved):
LGTM, thanks.
